### PR TITLE
feat(nvim_fetch): `vim.net.fetch()`

### PR DESCRIPTION
**Pull request by mrshmllow**
**draft**. a re-write of #23461, which got a bit out of hand. a pr implementing `:e URL` (and friends) will come after this one.

partially implements #23232.

### Todo
- [x] `user` opt
- [x] uploading files
- [ ] cmake stuff
- [ ] `redirect` opt
- [ ] write tests
- [ ] leave draft status

### Modivation

By having such a function built-in, plugins and users can have a standard way to download and fetching resources. This also allows us to move some editor commands such as `:edit` away from netrw.

### Reason for rewrite

In #23461, we decided it would be best to not rely on the `curl` command and instead use libcurl directly.

### Reasons for doing it ourselves

Lua-cURLv3 is looking quite old already, and in reality we are just offloading the burden of writing C to a different party.

Libcurl is quite easy, and anyone can write a wrapper. But this is not just any old wrapper. is quite specifically *neovim* code. not just `libcurl<->lua` code.

Curl provides its own solution for async transfers, the multi interface, but neovim has its own internal event system that multi doesnt quite fit into without further event loop modification.

Even in the multi-interface example on the Lua-cURLv3 repository it requires blocking to resolve the requests so we would need another wrapper to thread them or something and at that point we could introduce a nicer API, and at *that* point we could of written our own thing.

### CODE REVIEW

**Comment by clason** [comment]
I think we would still prefer to keep the API and (implementation) in Lua, similarly to `vim.mpack` and `vim.json`. The `nvim` API is pretty much exclusively for interfacing with the internal editor state.

**Comment by ii14** [comment]
I don't know if the implementation itself has to necessarily be in Lua, but yes, I don't think this should be a part of nvim API. You have a Lua `on_complete` callback, so that makes it already Lua only. See `nlua_state_init` function in `executor.c`, that's where `vim` global is initialized and you can add something like `vim.net` there

**Comment by lewis6991** [comment]
Not everything has to be done in C. Usually the heavy lifting is done in C with an internal binding (e.g. `vim._fetch()`), then an optional Lua wrapper can be added for a nicer interface.

**Comment by mrshmllow** [comment]
> Not everything has to be done in C. Usually the heavy lifting is done in C with an internal binding (e.g. vim._fetch()), then an optional Lua wrapper can be added for a nicer interface.

I really like this idea especially for things like accepting a table as data or `.json()`.

**Comment by famiu** [comment]
@mrshmllow Would you like me to help with the PR? I'd be happy to help if required..

**Comment by justinmk** [comment]
> In #23461, we decided it would be best to not rely on the `curl` command and instead use libcurl directly.

?? I haven't had a chance to read the discussion there. Why can't we start with `curl`? We can always change the implementation later, although adding another library to `nvim` doesn't sound great.

The main idea is:

1. start with a reasonable interface that can be extended in the future
2. address the _common case_ of "I want to download a file, ffs"

**Comment by clason** [comment]
>  Why can't we start by using the curl command? We can always change the implementation later, although adding another library to nvim doesn't sound great.

Yes, reading the discussion is necessary. TL;DR:
1. the `curl` command is argument hell; different versions (and platforms) have subtly (and not so subtly) different flags and behaviors;
2. the `libcurl` _library_, on the other hand, has a _very_ stable ABI, and a strong promise to keep it that way;
3. the result of that long discussion is _not_ to bundle `libcurl` (for now) and call into the **system** libcurl, which pretty much every supported platform ships with. (This will turn `curl` into a runtime dependency, much like `glibc`.)

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -3351,4 +3351,91 @@ totable({f}, {...})                                       *vim.iter.totable()*
...
+
+==============================================================================
+Lua module: net                                                      *lua-net*
+
+fetch({url}, {opts})                                         *vim.net.fetch()*
+    Perform an asyncronous network request.
+
+    Example: >lua
+    vim.net.fetch("https://example.com/api/get", {
+      on_complete = function (response)
```
Comment (justinmk) [inline runtime/doc/lua.txt:3363]:
```suggestion
      on_result = function (response)
```

or `on_done`. 

Looks like [nodejs](https://nodejs.org/api/net.html#netcreateconnectionoptions-connectlistener) calls this `on_end` , are there other conventions?

**Comment by justinmk** [comment]
> > Why can't we start by using the curl command? We can always change the implementation later, although adding another library to nvim doesn't sound great.

Ok, I read the discussion in https://github.com/neovim/neovim/pull/23461 and replied there.


> Yes, reading the discussion is necessary. TL;DR:
> 
> 1. the `curl` command is argument hell; different versions (and platforms) have subtly (and not so subtly) different flags and behaviors;

That doesn't matter and is premature optimization.

What matters right now is

1. eliminating netrw for "download" functionality
2. start with a simple, _synchronous_ interface that allows future revisions (async, "full curl parameters" or whatever...)

libcurl is not the right approach to start with.

**Pull request closed without merging**